File: week3/java_week3/Pen&Paper/Pen.java
public class Pen {
    private int inkAmount;
    private int inkCapacity;

    public Pen(int inkCapacity) {
        this.inkCapacity = inkCapacity;
        this.inkAmount = inkCapacity;
    }

    public Pen() {
        this(4096);
    }

    public int getInkAmount() {
        return inkAmount;
    }

    public int getInkCapacity() {
        return inkCapacity;
    }

    public void write(Paper paper, String message) throws OutOfInkException, OutOfSpaceException {
        if ( this.inkCapacity < 0 ) {
            System.out.println("Wrong pen capacity.");
            return;
        }

        if ( this.inkAmount == 0 ) {
            throw new OutOfInkException();
        }

        if ( this.inkAmount < message.length() ) {
            paper.addContent(message.substring(0, this.inkAmount));
            this.inkAmount = 0;
            
            return;
        }

        paper.addContent(message);
        this.inkAmount -= message.length();
    }

    public void refill() {
        if ( this.inkCapacity < 0 ) {
            System.out.println("Wrong pen capacity.");
            return;
        }
        this.inkAmount = this.inkCapacity;
    }
}

File: week3/cpp_hw_week3/Pen&Paper/Pen.cpp
#include "Paper.h"
#include "Pen.h"

Pen::Pen(int inkCapacity) {
    this->inkCapacity = inkCapacity;
    this->inkAmount = inkCapacity;
}

Pen::~Pen() {

}

int Pen::getInkAmount() const {
    return this->inkAmount;
}

int Pen::getInkCapacity() const {
    return this->inkCapacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    if ( this->inkCapacity < 0 ) {
        std::cout << "You have a wrong pen capacity." << std::endl;
        return;
    }

    if ( this->inkAmount == 0 ) {
        throw OutOfInk();
    }

    if ( this->inkAmount < message.length() ) {
        paper.addContent(message.substr(0, this->inkAmount));
        this->inkAmount = 0;

        return;
    }

    paper.addContent(message);
    this->inkAmount -= message.length();
}

void Pen::refill() {
    if ( this->inkCapacity < 0 ) {
        std::cout << "You have a wrong pen capacity. Change it." << std::endl;
        return;
    }
    this->inkAmount = this->inkCapacity;
}